### PR TITLE
Add support for the inbound-cidrs annotation

### LINKED ISSUE #294: Add support for more restrictive cidrs on alb-managed security groups
It would be good to add support similar to the port annotation that would allow the user to specify a list of inbound cidr ranges which they would like to restrict ingress to for the managed security groups.

```yml
apiVersion: extensions/v1beta1
kind: Ingress
metadata:
  name: my-ingress
  annotations:
    alb.ingress.kubernetes.io/scheme: internet-facing
    alb.ingress.kubernetes.io/subnets: subnet-1234,subnet-1235
    alb.ingress.kubernetes.io/listen-ports: '[{"HTTPS": 443}]'
    # cidrs to restrict ingress to
    alb.ingress.kubernetes.io/inbound-cidrs: 7.7.7.7/32,8.8.8.8/32
  labels:
    app: my-ingress
spec:
  rules:
  - host: test.com
    http:
      paths:
      - path: /path
        backend:
          serviceName: my-service
          servicePort: 443
```

We have a patch for this and will submit a pull request soon.

### PR DESCRIPTION
**Pull request by MarkFreebairn**
If the `alb.ingress.kubernetes.io/inbound-cidrs` annotation is set then the more restrictive CIDR IPv4 range(s) will be added (vs. the default `0.0.0.0/0`).

Fix #294.

I paired with @erran on these changes.

# Test cases
The following test cases were considered/run manually:

- [x] Starting with no annotation (inbound cidr defaults to 0.0.0.0/0 on managed sg) -> adding annotation adds new inbound cidr range and removes default 0.0.0.0/0
- [x] appending new cidr to annotation will add the new inbound source to the managed sg
- [x] removing a cidr from the annotation will remove it from the sg
- [x] removing the annotation entirely will remove all inbound cidrs and add the default 0.0.0.0/0
- [x] adding a new port will add all inbound cidrs from the annotation

### CODE REVIEW

**Comment by MarkFreebairn** [comment]
Hey @joshrosso would you mind having a look at this PR if you have some spare time please?

File: pkg/alb/loadbalancer/loadbalancer.go [hunk 1]
```diff
@@ -547,6 +555,14 @@ func (lb *LoadBalancer) needsModification() (loadBalancerChange, bool) {
 	}
 
 	if lb.CurrentPorts != nil && lb.CurrentManagedSG != nil {
+		if lb.CurrentInboundCidrs != nil {
```
Comment (bigkraig) [inline pkg/alb/loadbalancer/loadbalancer.go]: Not 100% sure if this is required. If CurrentInboundCidrs is nil but there is a desired inbound cidr, it won't return with true.

File: pkg/annotations/annotations.go [hunk 2]
```diff
@@ -33,6 +33,7 @@ const (
 	healthcheckTimeoutSecondsKey  = "alb.ingress.kubernetes.io/healthcheck-timeout-seconds"
 	healthyThresholdCountKey      = "alb.ingress.kubernetes.io/healthy-threshold-count"
 	unhealthyThresholdCountKey    = "alb.ingress.kubernetes.io/unhealthy-threshold-count"
+	inboundCidrsKey               = "alb.ingress.kubernetes.io/inbound-cidrs"
```
Comment (bigkraig) [inline pkg/annotations/annotations.go]: If `inbound-cidrs` is a configuration of the security group, should this be something like `security-group-inbound-cidrs` ?

File: pkg/aws/ec2/ec2.go [hunk 3]
```diff
@@ -264,24 +285,36 @@ func (e *EC2) UpdateSGIfNeeded(vpcID *string, sgName *string, currentPorts []int
...
 	}
 
+	if len(desiredCidrs) == 0 {
+		desiredCidrs = append(desiredCidrs, aws.String("0.0.0.0/0"))
+	}
+
 	// for each addPort, run an authorize to ensure it's added
 	for _, port := range desiredPorts {
-		if existsInOtherPortRange(port, currentPorts) {
+		ipranges := []*ec2.IpRange{}
```
Comment (bigkraig) [inline pkg/aws/ec2/ec2.go]: should this be ipRanges?

File: pkg/aws/ec2/ec2.go [hunk 4]
```diff
@@ -293,22 +326,29 @@ func (e *EC2) UpdateSGIfNeeded(vpcID *string, sgName *string, currentPorts []int
 
 	// for each currentPort, run a revoke to ensure it can be removed
 	for _, port := range currentPorts {
-		if existsInOtherPortRange(port, desiredPorts) {
+		ipranges := []*ec2.IpRange{}
```
Comment (bigkraig) [inline pkg/aws/ec2/ec2.go]: ipRanges

File: pkg/aws/ec2/ec2.go [hunk 5]
```diff
@@ -371,18 +420,24 @@ func (e *EC2) CreateSecurityGroupFromPorts(vpcID *string, sgName *string, ports
 		GroupId: oSG.GroupId,
 	}
 
+	if len(cidrs) == 0 {
+		cidrs = append(cidrs, aws.String("0.0.0.0/0"))
+	}
+
 	// for every port specified, allow all tcp traffic.
 	for _, port := range ports {
+		ipranges := []*ec2.IpRange{}
```
Comment (bigkraig) [inline pkg/aws/ec2/ec2.go]: ipRanges

**Comment by MarkFreebairn** [comment]
Hey @bigkraig thanks a lot for reviewing, sorry it took me so long to get back to looking at this! I have updated based on your comments, fixed merge conflicts and ran through all of the manual tests as well to make sure everything is still working correctly.

File: pkg/alb/loadbalancer/loadbalancer.go [hunk 6]
```diff
@@ -586,14 +586,12 @@ func (lb *LoadBalancer) needsModification() (loadBalancerChange, bool) {
...
-			if !reflect.DeepEqual(lb.DesiredInboundCidrs, lb.CurrentInboundCidrs) {
-				changes |= managedSecurityGroupsModified
-			}
+		sort.Sort(util.AWSStringSlice(lb.CurrentInboundCidrs))
+		sort.Sort(util.AWSStringSlice(lb.DesiredInboundCidrs))
+		if !reflect.DeepEqual(lb.DesiredInboundCidrs, lb.CurrentInboundCidrs) {
+			changes |= managedSecurityGroupsModified
 		}
-
+
```
[the remark this answers is not in the document]
Comment (bigkraig) [reply pkg/alb/loadbalancer/loadbalancer.go]: @MarkFreebairn sorry it took me until now for this minor request. apparently i had left it in pending for the last month

[the remark this answers is not in the document]
Comment (MarkFreebairn) [reply pkg/alb/loadbalancer/loadbalancer.go]: @bigkraig ah whoops sorry missed this! I've updated to fix merge conflicts and remove the additional whitespace now

**Pull request merged**